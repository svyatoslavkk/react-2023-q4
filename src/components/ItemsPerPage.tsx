import React from 'react';
import { useRouter } from 'next/router';
import { updateQueryParams } from '../functions/updateQueryParams';

function ItemsPerPage(props: Record<'pageSize', number>) {
  const { pageSize } = props;
  const router = useRouter();

  const optionValues = [10, 20, 30];

  return (
    <div className="items-per-page">
      <select
        name="itemsPerPage"
        id="itemsPerPage"
        data-testid="select"
        value={pageSize.toString()}
        onChange={(e) => {
          router.push(
            '?' + updateQueryParams(router.query, 'limit', e.target.value),
          );
        }}
      >
        {optionValues.map((el) => (
          <option key={el} value={el} data-testid="select-option">
            {el}
          </option>
        ))}
      </select>
    </div>
  );
}

export default ItemsPerPage;
