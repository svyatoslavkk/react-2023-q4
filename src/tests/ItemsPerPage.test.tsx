import React from 'react';
import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import mockRouter from 'next-router-mock';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';
import ItemsPerPage from '../components/ItemsPerPage';

vi.mock('next/router', () => vi.importActual('next-router-mock'));

describe('Tests for the Select component', () => {
  it('Value changes correctly', async () => {
    mockRouter.push('/?limit=10');
    render(<ItemsPerPage pageSize={Number(mockRouter.query.limit)} />, {
      wrapper: MemoryRouterProvider,
    });
    const select = screen.getByTestId('select');
    expect(select).toHaveValue('10');
    fireEvent.change(select, { target: { value: '5' } });
    expect(mockRouter.query).toEqual({ limit: '5' });
    cleanup();
    const limit = render(
      <ItemsPerPage pageSize={Number(mockRouter.query.limit)} />,
      {
        wrapper: MemoryRouterProvider,
      },
    );
    expect(screen.getByTestId('select')).toHaveValue('5');
    expect(limit.baseElement).toMatchSnapshot();
  });
});
