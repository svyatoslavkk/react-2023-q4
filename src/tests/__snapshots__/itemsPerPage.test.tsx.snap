// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for the Select component > Value changes correctly 1`] = `
<body>
  <div>
    <div
      class="items-per-page"
    >
      <select
        data-testid="select"
        id="itemsPerPage"
        name="itemsPerPage"
      >
        <option
          data-testid="select-option"
          value="10"
        >
          10
        </option>
        <option
          data-testid="select-option"
          value="20"
        >
          20
        </option>
        <option
          data-testid="select-option"
          value="30"
        >
          30
        </option>
      </select>
    </div>
  </div>
</body>
`;
